syntax = "proto3";

package org.couchers.api.core;

import "google/protobuf/timestamp.proto";
import "google/protobuf/wrappers.proto";
import "google/protobuf/empty.proto";

service API {
  // This is a secure service: a user needs to be authenticated and not jailed to call functions here, refer to auth.proto and jail.proto

  rpc Ping(PingReq) returns (PingRes) {
    // Pings the server for updates and basic user info
  }

  rpc GetUser(GetUserReq) returns (User) {
    // Get info about a particular user
  }

  rpc UpdateProfile(UpdateProfileReq) returns (google.protobuf.Empty) {
    // Update your profile
    // *Nullable can be set to null to clear the value. is_null defaults to false.
  }

  rpc SendFriendRequest(SendFriendRequestReq) returns (google.protobuf.Empty);

  rpc ListFriendRequests(google.protobuf.Empty) returns (ListFriendRequestsRes);

  rpc ListFriends(google.protobuf.Empty) returns (ListFriendsRes);

  rpc RespondFriendRequest(RespondFriendRequestReq) returns (google.protobuf.Empty);

  rpc CancelFriendRequest(CancelFriendRequestReq) returns (google.protobuf.Empty);

  rpc Search(SearchReq) returns (SearchRes) {
    // Search for user names and usernames (for now)
  }

  // Report another user to admin. Raises INVALID_ARGUMENT if you
  // report yourself or NOT_FOUND if the reported user doesn't exist.
  rpc Report(ReportReq) returns (google.protobuf.Empty);

  // Filter references based on from user and type, and
  // paginate the results according to client's wish.
  rpc GetGivenReferences(GetGivenReferencesReq) returns (GetReferencesRes);

  // Filter references based on to user or type, and
  // paginate the results according to client's wish.
  rpc GetReceivedReferences(GetReceivedReferencesReq) returns (GetReferencesRes);

  // Return the reference types that is, at the moment, possible to write to a user.
  rpc AvailableWriteReferenceTypes(AvailableWriteReferenceTypesReq) returns (AvailableWriteReferenceTypesRes);

  // Write a reference for someone. Raises INVALID_ARGUMENT if you
  // try to write for yourself, or NOT_FOUND if the user doesn't exist.
  // Raises FAILED_PRECONDITION if you try to write a reference twice.
  rpc WriteReference(WriteReferenceReq) returns (google.protobuf.Empty);

  rpc InitiateMediaUpload(google.protobuf.Empty) returns (InitiateMediaUploadRes) {
    // Initiate a media upload
  }
}

message SearchReq {
  string query = 1;
}

message SearchRes {
  repeated User users = 1;
}

message SendFriendRequestReq {
  uint64 user_id = 1;
}

message FriendRequest {
  uint64 friend_request_id = 1;

  enum FriendRequestStatus {
    // these are the only two states that a user will see
    PENDING = 0;
    ACCEPTED = 1;
  }

  FriendRequestStatus state = 2;

  uint64 user_id = 3;
}

message RespondFriendRequestReq {
  uint64 friend_request_id = 1;

  bool accept = 2;
}

message CancelFriendRequestReq {
  uint64 friend_request_id = 1;
}

message ListFriendRequestsRes {
  repeated FriendRequest sent = 1;
  repeated FriendRequest received = 2;
}

message ListFriendsRes {
  repeated uint64 user_ids = 1;
}

message PingReq {}

message PingRes {
  User user = 1;
  uint32 unseen_message_count = 2;
  uint32 unseen_host_request_count = 3;
  uint32 pending_friend_request_count = 4;
}

message MutualFriend {
  uint64 user_id = 1;
  string username = 2;
  string name = 3;
}

enum HostingStatus {
  HOSTING_STATUS_UNSPECIFIED = 0;
  HOSTING_STATUS_UNKNOWN = 1;
  HOSTING_STATUS_CAN_HOST = 2;
  HOSTING_STATUS_MAYBE = 3;
  HOSTING_STATUS_DIFFICULT = 4;
  HOSTING_STATUS_CANT_HOST = 5;
}

enum SmokingLocation {
  SMOKING_LOCATION_UNSPECIFIED = 0;
  SMOKING_LOCATION_UNKNOWN = 1;
  SMOKING_LOCATION_YES = 2;
  SMOKING_LOCATION_WINDOW = 3;
  SMOKING_LOCATION_OUTSIDE = 4;
  SMOKING_LOCATION_NO = 5;
}

message NullableUInt32Value {
  bool is_null = 1;
  uint32 value = 2;
}

message NullableBoolValue {
  bool is_null = 1;
  bool value = 2;
}

message NullableStringValue {
  bool is_null = 1;
  string value = 2;
}

message User {
  enum FriendshipStatus {
    NOT_FRIENDS = 0;
    FRIENDS = 1;
    PENDING = 2;
    NA = 3; // not applicable, e.g. user requesting is self
  }

  uint64 user_id = 1;
  string username = 2;
  string name = 3;
  string city = 4;

  // doubles are enough to describe lat/lng
  // in EPSG4326, lat & lon are in degress, check the EPS4326 definition for details
  double lat = 33;
  double lng = 34;
  double radius = 35; // meters

  double verification = 5;
  double community_standing = 6;
  uint32 num_references = 7;
  string gender = 8;
  uint32 age = 9;
  string color = 10; // web color, e.g. #643073, case doesn't matter
  google.protobuf.Timestamp joined = 11; // not exact
  google.protobuf.Timestamp last_active = 12; // not exact
  HostingStatus hosting_status = 13;
  string occupation = 14;
  string about_me = 15;
  string about_place = 16;
  repeated string languages = 17;
  repeated string countries_visited = 18;
  repeated string countries_lived = 19;

  FriendshipStatus friends = 20;
  repeated MutualFriend mutual_friends = 21;

  google.protobuf.UInt32Value max_guests = 22;
  google.protobuf.BoolValue multiple_groups = 23;
  google.protobuf.BoolValue last_minute = 24;
  google.protobuf.BoolValue accepts_pets = 25;
  google.protobuf.BoolValue accepts_kids = 26;
  google.protobuf.BoolValue wheelchair_accessible = 27;
  SmokingLocation smoking_allowed = 28;
  google.protobuf.StringValue sleeping_arrangement = 29;
  google.protobuf.StringValue area = 30;
  google.protobuf.StringValue house_rules = 31;

  string avatar_url = 32;
}

message GetUserReq {
  string user = 1;
}

message RepeatedStringValue {
  // A RepeatedStringValue is like a "protobuf wrapper" (like {Type}Value) but for repeated strings
  // allows us to determine if the user wanted to remote this list of strings
  bool exists = 1;
  repeated string value = 2;
}

message UpdateProfileReq {
  // These StringValue types allow us to determine if the value is absent versus ""

  google.protobuf.StringValue name = 1;

  google.protobuf.StringValue city = 2;
  google.protobuf.DoubleValue lat = 22;
  google.protobuf.DoubleValue lng = 23;
  google.protobuf.DoubleValue radius = 24;

  google.protobuf.StringValue gender = 3;
  NullableStringValue occupation = 4;
  NullableStringValue about_me = 5;
  NullableStringValue about_place = 6;
  google.protobuf.StringValue color = 7;

  HostingStatus hosting_status = 8;

  RepeatedStringValue languages = 9;
  RepeatedStringValue countries_visited = 10;
  RepeatedStringValue countries_lived = 11;

  NullableUInt32Value max_guests = 12;
  NullableBoolValue multiple_groups = 13;
  NullableBoolValue last_minute = 14;
  NullableBoolValue accepts_pets = 15;
  NullableBoolValue accepts_kids = 16;
  NullableBoolValue wheelchair_accessible = 17;
  SmokingLocation smoking_allowed = 18;
  NullableStringValue sleeping_arrangement = 19;
  NullableStringValue area = 20;
  NullableStringValue house_rules = 21;
}

message ReportReq {
  uint64 reported_user_id = 1;

  // one-line summary
  string reason = 2;

  // multi-line free text
  string description = 3;
}

enum ReferenceType {
  FRIEND = 0;  // We met on an event but haven't surfed or hosted
  SURFED = 1;  // The author have surfed, the referent has hosted
  HOSTED = 2;  // The author have hosted, the referent has surfed
}

message OptionalReferenceType {
  ReferenceType value = 1;
}

message GetGivenReferencesReq {
  uint64 from_user_id = 2;
  OptionalReferenceType type_filter = 3;                // optional

  // Pagination; default is to get the N latest references that matches the filter,
  // skipping the start_at most recent ones.
  uint32 number = 5;
  uint32 start_at = 4;
}

message GetReceivedReferencesReq {
  uint64 to_user_id = 2;
  OptionalReferenceType type_filter = 3;

  // Pagination; default is to get the N latest references that matches the filter,
  // skipping the start_at most recent ones.
  uint32 number = 5;
  uint32 start_at = 4;
}

message Reference {
  uint64 from_user_id = 1;
  uint64 to_user_id = 2;
  ReferenceType reference_type = 3;
  string text = 4;
  google.protobuf.Timestamp written_time = 5;  // not exact
}

message GetReferencesRes {
  // Total number of records that matches the filter
  uint32 total_matches = 1;

  // A list of zero to `number` references
  repeated Reference references = 2;
}

message AvailableWriteReferenceTypesReq {
  uint64 to_user_id = 1;
}

message AvailableWriteReferenceTypesRes {
  repeated ReferenceType reference_types = 1;
}

message WriteReferenceReq {
  uint64 to_user_id = 1;
  ReferenceType reference_type = 2;
  string text = 3;
  bool was_safe = 4;
  int32 rating = 5;
}

message InitiateMediaUploadRes {
  // URL to POST the data to
  string upload_url = 1;
  google.protobuf.Timestamp expiry = 2;
}
